### ISSUE TITLE
Is there a way to reset the consumer group offsets to latest

### ISSUE THREAD

**Issue by chandradeepak** [bsm/sarama-cluster#87]
In the sarama-cluster is there a easy way to reset the consumer group offsets. In old kafka all you had to do is delete the /consumers/{consumername} tree tierarchy recursively. you are done . all the offsets related to that consumer group are gone.
But with new kafka 0.9  and more the consumer group offsets are maintained in kafka only in the topic __consumer_offsets. so how can we reset the offsets of a consumer group specifically so that i can start reading from latest offset . 

Is there a way to do that in sarama-cluster ?

May be having a new function called ResetOffsets which would read the offsets for all topics and partitions for that consumer group and mark the latest offsets. Just like how it does during the start.  May be that would work.

**Comment by elakito** [comment]
Hi,
I looked at the sarama-cluster code and it seems there is no mechanism to change the offset from the current committed offset. 

So, I experimented with the idea of adding Seek method to Consumer and also exposing the initial rebalance notification to the consumer so that it can update the offset value to its desired value before the subscription starts. This approach seems to work. But I don't know if this is the right approach. So, I'll share my code and ask for any comments.
regards, aki

**Comment by elakito** [comment]
Hi,
My change is available at my forked repo's work-87 branch.
https://github.com/elakito/sarama-cluster/tree/work-87
consumer.go includes the change mentioned above and consumer_test.go includes a test that has a code fragment (see the link below) to retrieve the starting offsets from elsewhere and use them to subscribe to the assigned/claimed partitions instead of using the default offset values.

https://github.com/elakito/sarama-cluster/blob/work-87/consumer_test.go#L210-L221

I would appreciate if sarama-cluster developers can comment on this approach.
thanks.
regards, aki

**Comment by dim** [comment]
@elakito I am a little confused about the approach, can you walk me through? Assuming you want to rewind to previous offsets, you call `Seek` for all topic partition combinations you find, but how do you then restart the consumer/trigger a rebalance?

**Comment by elakito** [comment]
@dim thank you for reviewing the code. 
The intentional use case of Seek is to reposition the offset after a rebalancing takes place and the partition is assigned to that consumer instance. Thus, the method is intended to be used within the rebalance notification processing and it does not actively trigger a rebalancing.

regards, aki

**Comment by dim** [comment]
Sorry, but it's really hard to see the purpose here without a more elaborate use case. You are consuming a feed, committing offsets but then set an internal marker to a previous offset to resume from after rebalancing? Your `Seek` call only stores the resume-position in local state. What if that partition is then assigned to another consumer? 

In it's current state, it simply doesn't make sense to me. What am I missing?

**Comment by elakito** [comment]
@dim one of the use cases is to store the offset at the data consuming component in a transaction. In this case, the committed offsets at the brokers may not match exactly to the actually consumed offsets. 

Concretely, in this case, when the consumer crashes and restarts, it can restart from the offset stored at this shared store, independently from the committed offset at the broker (either the offsets at the broker are not committed at all or updated periodically only to show the progress, etc). And the currently active consumer is supposed to be updating the offset value at the store while it consumes the messages, this offset value can be used by another consumer when it picks up the consuming role. 

To pass this offset value to the consumer instance to be used as its starting/resuming offset, I am storing it locally.

Is the use case clear now?

I thought this idea is analogue to kafka Consumer's seek method (except its use is more flexible due to its synchronous poll() behavior)  and this would be reasonable.
https://kafka.apache.org/0100/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#seek(org.apache.kafka.common.TopicPartition,%20long)

However, I am open to alternative approaches.

Thanks.
regards, aki

**Comment by chandradeepak** [comment]
@dim , my use case is more of if i start a consumer group and then after some time , client want to consume the data from latest offset instead of the offset stored in kafka. One way do that is i can start a new consumer group with different subscriber name. But if we have this client connecting often and disconnecting we might end up having too many consumer groups with out being deleted until the retention period is triggered. 

Please let me know your thoughts.

**Comment by panamafrancis** [comment]
@dim My use case is during a test run we produce n messages to kafka then abort the test, without creating new consumer groups per test or setting the retention period very low, the consumers would be started at the last offset for their assigned partition, which could mean waiting for all the aborted runs messages to be consumed. 

So our simple consumers can be reset to the latest offset fairly easily and i'd like to find a similar solution here.

@elakito I think your solution is on the right track but i agree with @dim in that it does look a bit broken. Should you not at least mark the topic as dirty?

**Comment by elakito** [comment]
@panamafrancis thank you for your comments. As mentioned in my comment from Dec 6, my original intention was to have the Seek method to be used only during the rebalancing phase when a new subscription is created for the claimed partition. My use case is covered by this usage and I believe so are @chandradeepak and your use cases. My initial code, however, contained the code that affected the existing subscription's state, which raised concern.

So, I commented out the corresponding part from my code (also adding dirty = true in that part for the future consideration to handle this case).
https://github.com/elakito/sarama-cluster/blob/work-87/consumer.go#L184-L186

Does this clear your concern? 
@dim What do yo think? I think adding Seek method itself is reasonable. We can start with the current approach to support these use cases. If we find a better approach, we can switch the implementation of it, no?

regards, aki

**Comment by panamafrancis** [comment]
@elakito @chandradeepak 
I've decided to use a "workaround" whereby i shutdown all the consumer instances and create a new set of instances with a different consumer group name and the offset.initial == newest 

Maybe that's good enough for you both.

**Comment by elakito** [comment]
@panamafrancis As mentioned earlier there are two distinct use cases
1. forget about the previous offset and just start from the latest offset because you don't mind about losing messages.
2. start exactly from the offset that was processed and committed at the backend and do not lose messages. This offset value committed at the backend  may be different from the actual committed offset at the broker became the previous offset commit to the broker didn't reach the broker.
My use case is case 2 and my patch is intended to support both use cases, whereas your workaround is just a workaround for case 1.

**Comment by dim** [comment]
There are a few solutions for this:

1. Marked offsets have a [RetentionTime](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-OffsetCommitRequest) associated with them. If set to `0` it will effectively reset the stored offsets for the topic/partition. Retention times are configurable in https://godoc.org/github.com/Shopify/sarama#Config
2. I am not 100% sure why you would want to "forget about the previous offset", but you could always query the latest known offset for a topic/partition via `client.GetOffset(topic, partition, sarama.OffsetNewest)` and the resulting value for [consumer.MarkPartitionOffset](https://godoc.org/github.com/bsm/sarama-cluster#Consumer.MarkPartitionOffset) and restart the consumer.

**Comment by chandradeepak** [comment]
Hello Dim,
i have a question here. In the documentation it says if the retention is set to 0 it is disabled. So how does setting to 0 have any effect . 

Also when you say restart subscriber does that mean we need to do consumer.Close and then start the subscription again with same topics. 

can you please clarify ?





/ The retention duration for committed offsets. If zero, disabled
            // (in which case the `offsets.retention.minutes` option on the
            // broker will be used).  Kafka only supports precision up to
            // milliseconds; nanoseconds will be truncated. Requires Kafka
            // broker version 0.9.0 or later.
            // (default is 0: disabled).

**Comment by dim** [comment]
So I am not 100% sure the above is correct, but that's something you need to bring up with the Sarama maintainers. According to Kafka's documentation, you need to set the setting to `-1` to disable it (= persist indefinitely), while `0` is a valid setting (= expire immediately).

**Comment by wyndhblb** [comment]
I'm gonna +1 this, a common use cases for streaming + aggregation process with "checkpoints/committed" messages.  Basically emit a message like [key, checkpoint offset] message "somewhere".  When things rebalance or restart, consumers need to be able to pick up where the last process left off, for given key, but that also means they need to rewind the topic to get to the "earliest" offset that has a check point for all the keys in the stream. 

While another consumer group can solve the issue (basically re-ingest the entire stream ignoring things until things hit the current offset checkpoint for a given key) that can be a long process depending on the topic retention and volume, and the rebalance on the same consumer group won't be able to reset itself, without some mechanism to set offsets for that partition back to the 0, or some other offset gotten from somewhere else.  This is a standard process in Spark's Structured streaming or Kafka Streams.

**Comment by nsd20463** [comment]
@wyndhblb if you s/key/partition/ this is what kafka consumer groups do. Per-partition checkpoints are efficient to do in kafka, where everything organized is per-partition. Anything further would be the responsibility of the consumer code.

**Comment by wyndhblb** [comment]
correct, there's currently no ability to re/set an offset for a partition in this lib.  This is not an issue of "where to store"  those checkpoints (there are many ways). Just ability to set them on either notification events or on initialization.  May just be as simple as exposing the ```ConsumePartition(topic string, partition int32, offset int64) ``` from sarama to be public.

**Comment by marcosrmendezthd** [comment]
would appreciate a way to change the kafka offsets. sometimes we have to backfill (process old stuff) or skip ahead. i tried the following after stopping the consumer, but it didn't work either...

```
cluster, clusterError := cluster.NewConsumer(
                 kafkaBrokers, kafkaConsumerGroup,[]string{kafkaTopic}, nil)
if clusterError != nil {
	return fmt.Errorf("Unable to connect to kafka brokers (consumer). %v", clusterError)
}

defer cluster.Close()

for partition, offset := range offsets {
    cluster.MarkPartitionOffset(kafkaTopic, partition, offset, "")
}

cluster.CommitOffsets()

```

**Comment by dim** [comment]
@wyndhblb @marcosrmendezthd sorry for the delayed response, it has been quite busy here. 

While I understand the intention, I am not so sure how this would work in practice. Let's assume the following, simple use case:

1. you are consuming a topic with two partitions (P0,P1) 
2. you have two consumers (A and B)
3. you have consumed 100 messages on each partition (200 messages in total)
4. now, you want to "rewind" to offset=50 on both partitions

I am not sure how you would do that? If you reset the partition offsets of `P0` and `P1` to 50 on `A`, you need to trigger a rebalance and make sure that `B` doesn't overwrite your commit, but that requires you to maintain some sort of shared state between your consumers and that's certainly out of scope for this library.

Thoughts?

**Comment by marcosrmendezthd** [comment]
@dim agreed. the scope of the state is outside the library. however, we do that regularly as we operate a production kafka cluster. we **stop** consumers and rewind based on a timestamp which is inside our messages and each partition gets the offset they need. sometimes we just want to move ahead and just set them to the newest offset.

all this is currently done with zookeeper-based offsets. however i'd like to migrate to kafka-based offsets if possible. so i need to be able to set them properly; -1, -2, specific offset.

does this make sense? :smile:

**Comment by dim** [comment]
So assuming you know exactly the topics, partitions and corresponding offsets for each, why don't you:

1. stop the consumers
2. use the `client.Coordinator(consumerGroup string) (*Broker, error)` to retrieve the coordinator
3. use https://godoc.org/github.com/Shopify/sarama#Broker.CommitOffset to update all offsets
4. restart consumers

The reason why I (the lib :)) cannot do this for you is the fact that it requires some sort of coordination between your consumer instances (as mentioned) and it's not something I can easily bake in.

Makes sense?

**Comment by marcosrmendezthd** [comment]
@dim i posted similar code in this thread maybe a month ago. i'm not sure why the commits are not saving properly. maybe some setting i'm missing? that's what i have to investigate. anything you can think of? that should work in all cases without additional config changes?

**Comment by dim** [comment]
Right, I see, there are two problems with your code. 

For once, `MarkPartitionOffset` only accepts values that are greater than the current offset. This is a limitation by design, to avoid a specific edge case race condition. 

Additionally, you cannot use a cluster consumer to rewind your offsets. As mentioned earlier, you might have two (or many more) consumers running in parallel and while one might reset all the offsets, the others will be unaware of the change and will simply overwrite/bump them up again in the next commit cycle.

For now, I can only see only one sensible solution. You stop all your consumers and go "low level": use client.Coordinator to fetch the broker and then broker.CommitOffsets.

**Comment by dim** [comment]
I will think of a proper solution for your problem in the meantime. The "right" process for solving this should be:

1. the cluster is consuming N topics/partition 
2. you join the cluster with a special "maintenance" consumer `M`, forcing a rebalance
3. the cluster is set up with a special partitioning strategy which detects `M` and assigns *all* partitions to it (i.e. all the other, normal consumers suddenly become idle)
4. `M` then (re)sets the offsets for all partitions
5. `M` leaves the cluster again forcing a new rebalance
6. partitions are re-assigned to normal consumers, these resume at newly stored offsets

**Comment by disq** [comment]
sounds good, but also complicated... in the meantime, if there's a backfill operation one could stop all consumer instances and set offsets with an external tool like [fgeller/kt](https://github.com/fgeller/kt) I think.

**Comment by dim** [comment]
Yep, it seems this can be solved by a maintenance/console tool and should not be part of the core consumer. Will close for now.

**Comment by chandradeepak** [comment]
Hello Dim, one question when you say 

"For now, I can only see only one sensible solution. You stop all your consumers and go "low level": use client.Coordinator to fetch the broker and then broker.CommitOffsets."

say for suppose the topic has 5 partitions and each are at 10th offset now that means we read 10 messages for each of the partition and total we read 50 messages .  now if i commit offset to 4 will the new offset be replayed from 4th offset onwards. Is that possible . 

i checked the __consumer_offsets topic . where the offsets are really stored .  this is how it looks like.
[topicA_consumer1,topicA,1]::[OffsetMetadata[10,NO_METADATA],CommitTime 1501025782054,ExpirationTime 1501112182054]
so for topicA the offset is at 10. if i commit offste now at 4.  will this be 
[topicA_consumer1,topicA,1]::[OffsetMetadata[4,NO_METADATA],CommitTime 1501025782054,ExpirationTime 1501112182054]

so does kafka allow writing offsets to older ones for a particular partition. so that it can replay. is that how the offset cache for  __consumer_offsets topic work ?

**Comment by dim** [comment]
> so does kafka allow writing offsets to older ones for a particular partition. so that it can replay. is that how the offset cache for __consumer_offsets topic work ?

Sorry, just checking if I understand this correctly, you're asking if you can force-commit an offset that is older than one you have committed earlier? And if you can, which of the two offsets would Kafka use when it resumes after a rebalance?

**Comment by dim** [comment]
Just confirmed, I added the following test case:

```
It("should allow to rewind offsets", func() {
	cmd := testCmd(
		testDataDir(testKafkaRoot, "bin", "kafka-topics.sh"),
		"--create",
		"--zookeeper", "127.0.0.1:22181",
		"--replication-factor", "1",
		"--partitions", "1",
		"--topic", "topic-u",
	)
	Expect(cmd.Run()).To(Succeed())

	Expect(testSeed(100, "topic-u")).To(Succeed())

	// connect and mark an offset
	cs, err := newConsumerOf(testGroup, "topic-u")
	Expect(err).NotTo(HaveOccurred())
	subscriptionsOf(cs).Should(HaveKeyWithValue("topic-u", HaveLen(1)))
	cs.MarkPartitionOffset("topic-u", 0, 30, "")
	Expect(cs.Close()).To(Succeed())

	// connect and consume next message then commit earlier offset
	cs, err = newConsumerOf(testGroup, "topic-u")
	Expect(err).NotTo(HaveOccurred())
	subscriptionsOf(cs).Should(HaveKeyWithValue("topic-u", HaveLen(1)))
	Expect(<-cs.Messages()).To(Equal(&sarama.ConsumerMessage{Value: []byte("PLAINDATA-00000031"), Topic: "topic-u", Offset: 31}))
	cs.MarkPartitionOffset("topic-u", 0, 20, "")
	Expect(cs.Close()).To(Succeed())

	// connect and consume next message
	cs, err = newConsumerOf(testGroup, "topic-u")
	Expect(err).NotTo(HaveOccurred())
	defer cs.Close()
	subscriptionsOf(cs).Should(HaveKeyWithValue("topic-u", HaveLen(1)))
	Expect(<-cs.Messages()).To(Equal(&sarama.ConsumerMessage{Value: []byte("PLAINDATA-00000021"), Topic: "topic-u", Offset: 21}))
})
```

It fails with:

```
  Expected
      <*sarama.ConsumerMessage | 0xc420025c70>: {
          Key: nil,
          Value: "PLAINDATA-00000031",
          Topic: "topic-u",
          Partition: 0,
          Offset: 31,
          Timestamp: 0001-01-01T00:00:00Z,
      }
  to equal
      <*sarama.ConsumerMessage | 0xc4200b03f0>: {
          Key: nil,
          Value: "PLAINDATA-00000021",
          Topic: "topic-u",
          Partition: 0,
          Offset: 21,
          Timestamp: 0001-01-01T00:00:00Z,
      }
```

**Comment by chandradeepak** [comment]
@dim , makes sense so it can't commit older offsets. thanks for the confirmation .

**Comment by dim** [comment]
@chandradeepak actually, it works when I remove the https://github.com/bsm/sarama-cluster/blob/master/partitions.go#L114 condition. Unfortunately, I am a bit hesitant about removing it. The problem is that would be easily misunderstood.

Assuming you are consuming a topic/partition and decide to "rewind", you would first need to call `MarkPartitionOffset` with the offsets you want to resume from and you would *have to* trigger a rebalance next. I don't see how this would work in practice, but ideas are welcome.

I could add a `RewindPartitionOffset(topic string, partition int32, offset int64, metadata string) error` method but it would only ever work if the consumer owns that topic/partition, which isn't probably what most people need either which brings me back to https://github.com/bsm/sarama-cluster/issues/87#issuecomment-286227486

**Comment by tygrash** [comment]
You can use `https://github.com/tygrash/ofter` to reset the offset to a specific value.
